server/worker/tasks.py

```diff
@@ -133,22 +149,60 @@ def run_task(task: BackgroundTask, db_session):
 
 
 def claim_next_task(worker_id: str, db_session) -> Optional[BackgroundTask]:
+    # Use SELECT ... FOR UPDATE to lock all queued tasks, ensuring that only a
+    # single worker can claim a single task at a time. This ensures that we get
+    # an accurate view of which tasks are available to claim. If another worker has
+    # already locked the queued task, this query will block (though it shouldn't
+    # take long to claim a task, so this shouldn't cause problems).
+    #
+    # If new tasks are added after we lock the queued tasks, they won't be
+    # accessible by this worker, but another worker also won't be able to claim
+    # them (since it will also be trying to lock _all_ queued tasks, including
+    # the ones we already locked).
+    #
+    # Importantly, this also ensures that we get a safe view of which lock_keys
+    # are currently in use by in-progress tasks. If we're the only worker that
+    # can claim a task right now, then no additional lock_keys can be locked. It
+    # is possible for another worker to finish an in-progress task and release a
+    # lock_key while we're running this query, but that's fine because even if
+    # we don't see that change and wanted to claim that lock_key, we'll just fail
+    # and try again later.
```

Comment:
I'm following the high-level, just struggling a bit to parse this third paragraph, specifically this sentence:

> If we're the only worker that can claim a task right now, then no additional lock_keys can be locked.

Reply:
Is this any clearer?

> Importantly, this also ensures that we get a safe view of which lock_keys are currently in use by in-progress tasks. If this worker has successfully locked all queued tasks, then it's the only worker that can claim a task right now, so no additional tasks can be started. That means the current set of in-progress lock_keys won't change until this worker releases the lock on the queued tasks.

Reply:
Yes that does read clearer, thanks for editing!